extensions/rvadapter/src/main/java/com/google/android/agera/rvadapter/RepositoryAdapter.java

```diff
@@ -318,13 +337,51 @@ public final void startObserving() {
    */
   public final void stopObserving() {
     observable.removeUpdatable(this);
+    updateAllObservable.removeUpdatable(updateAllUpdatable);
   }
 
   /**
-   * Invalidates the data set so {@link RecyclerView} will schedule a rebind of all data.
+   * Invalidates the data set so {@link RecyclerView} will schedule rebinds of changed data.
    */
   @Override
   public final void update() {
+    if (dataInvalid) {
+      return;
+    }
+
+    int totalItemsAddedCount = 0;
+    for (int i = 0; i < repositoryCount; i++) {
+      endPositions[i] += totalItemsAddedCount;
+
+      final Object newData = repositories[i].get();
+      if (!newData.equals(data[i])) {
+        data[i] = newData;
+        final int itemCount = presenters[i].getItemCount(data[i]);
+        final int startPosition = i > 0 ? endPositions[i-1] : 0;
+        final int endPosition = startPosition + itemCount;
+        final int itemsAddedCount = endPosition - endPositions[i];
+        if (itemsAddedCount == 0) {
+          notifyItemRangeChanged(startPosition, itemCount);
+        } else {
+          totalItemsAddedCount += itemsAddedCount;
+          endPositions[i] += itemsAddedCount;
+
+          if (itemsAddedCount > 0) {
+            notifyItemRangeChanged(startPosition, itemCount - itemsAddedCount);
```

Comment:
Multiple problems I can think of:

- in CompiledRepository at least, I'm not preventing the client code to reuse the same value object for new data, because the `.notifyIf()` configuration lets the user do their own judgement about whether to fire the update event. Then in here, the assumption that `!newData.equals(data[i])` means a change has happened is invalid.

- Not too sure about the blanket assumption that `itemRangeChanged` + `itemRangeInserted`/`itemRangeRemoved` is adequate for all cases. Some clients might want to tweak this (e.g. use the shiny new [DiffUtil](https://developer.android.com/reference/android/support/v7/util/DiffUtil.html) to do all the event firing correctly). I would probably just let the `notifyDataSetChanged()` bit be overridable, so client code can do all sorts of fancy fine-grained event firing.